src/kernels/jupyter/finder/universalRemoteKernelFinder.ts

```diff
@@ -0,0 +1,397 @@
+            // Its possible the connection is dead, just display the live kernels we had.
+            // I.e. if user had a notebook connected to a remote kernel, then just display that live kernel.
+            kernels = await this.getFromCache(updateCacheCancellationToken.token);
+            kernels = kernels.filter((item) => item.kind === 'connectToLiveRemoteKernel');
+        }
+
+        if (updateCacheCancellationToken.token.isCancellationRequested) {
+            return;
+        }
+
+        await this.writeToCache(kernels);
+
+        this._onDidChangeKernels.fire();
+    }
+
+    /**
+     *
+     * Remote kernel finder is resource agnostic.
+     */
+    listContributedKernels(_resource: Resource): KernelConnectionMetadata[] {
+        return this.cache;
+    }
+
+    private async getRemoteConnectionInfo(
+        cancelToken?: CancellationToken
+    ): Promise<INotebookProviderConnection | undefined> {
+        const ui = new DisplayOptions(false);
+        return this.notebookProvider.connect({
+            resource: undefined,
+            ui,
+            localJupyter: false,
+            token: cancelToken,
+            serverId: this.serverUri.serverId
+        });
+    }
+
+    private async getFromCache(cancelToken?: CancellationToken): Promise<RemoteKernelConnectionMetadata[]> {
+        let results: RemoteKernelConnectionMetadata[] = this.cache;
+        const key = this.getCacheKey();
+
+        // If not in memory, check memento
+        if (!results || results.length === 0) {
+            // Check memento too
+            const values = this.globalState.get<{
+                kernels: RemoteKernelConnectionMetadata[];
+                extensionVersion: string;
+            }>(key, { kernels: [], extensionVersion: '' });
+
+            if (values && isArray(values.kernels) && values.extensionVersion === this.env.extensionVersion) {
+                results = values.kernels.map(deserializeKernelConnection) as RemoteKernelConnectionMetadata[];
+                this.cache = results;
+            }
+        }
+
+        // Validate
+        const validValues: RemoteKernelConnectionMetadata[] = [];
+        const promise = Promise.all(
+            results.map(async (item) => {
+                if (await this.isValidCachedKernel(item)) {
+                    validValues.push(item);
+                }
+            })
+        );
+        if (cancelToken) {
+            await Promise.race([
+                promise,
+                createPromiseFromCancellation({ token: cancelToken, cancelAction: 'resolve', defaultValue: undefined })
+            ]);
+        } else {
+            await promise;
+        }
+        return validValues;
+    }
+
+    // Talk to the remote server to determine sessions
+    @capturePerfTelemetry(Telemetry.KernelListingPerf, { kind: 'remote' })
+    public async listKernelsFromConnection(
+        connInfo: INotebookProviderConnection
```

Comment:
i think we can pass the cancellation token in here, there's a lot of async called in here